### PR TITLE
refactor vscode-ext to be more extensible

### PR DESCRIPTION
**Pull request by infiniteregrets**

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -228,7 +228,7 @@ jobs:
   test_agent_image:
     runs-on: ubuntu-latest
     needs: changed_files
-    if: ${{needs.changed_files.outputs.rs_changed == 'true' || needs.changed_files.outputs.ci_changed == 'true'}}
+    if: ${{needs.changed_files.outputs.rs_changed == 'true' || needs.changed_files.outputs.ci_changed == 'true' || needs.changed_files.outputs.vscode_changed == 'true'}}
```
Comment (eyalb181) [inline .github/workflows/ci.yaml:231]: Why is this necessary?

File: vscode-ext/src/util.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import * as vscode from 'vscode';
+
+/// Key used to store the last selected target in the persistent state.
+export const LAST_TARGET_KEY = "mirrord-last-target";
+
+/// Option in the target selector that represents no target.
+export const TARGETLESS_TARGET_NAME = "No Target (\"targetless\")";
+
+// Display error message with help
+export function mirrordFailure(err: any) {
```
Comment (eyalb181) [inline vscode-ext/src/util.ts:10]: I'd merge this entire file into api.ts, it's only used there.

File: vscode-ext/src/debugger.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+			return config;
+		}
+
+		// For some reason resolveDebugConfiguration runs twice for Node projects. __parentId is populated.
+		if (config.__parentId || config.env?.["__MIRRORD_EXT_INJECTED"] === 'true') {
+			return config;
+		}
+
+		// do not send telemetry for CI runs
+		if (vscode.env.isTelemetryEnabled && !CI_BUILD_PLUGIN) {
```
Comment (eyalb181) [inline vscode-ext/src/debugger.ts:50]: The checkVersion logic is split between this file and versionCheck.ts. Can this entire `if` be moved into versionCheck.ts?

**Review by eyalb181 — CHANGES_REQUESTED** [review]
Looks good in general, made some comments

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (infiniteregrets) [reply .github/workflows/ci.yaml:231]: thanks for the review! the `vscode_e2e` job depends on `test_agent_image` and `build_mirrord` and hence they need to run as well to make `vscode_e2e` run

File: vscode-ext/src/util.ts (hunk 2 above)
Comment (infiniteregrets) [reply vscode-ext/src/util.ts]: its also used by the extension.ts file, however i believe it makes sense to keep it with the API, we don't really have any "utility functions" as such anyway, fixed!

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (eyalb181) [reply .github/workflows/ci.yaml:231]: I understand that, but why is the vscode job dependent on the agent tests running (as opposed to just building the agent image)?

File: vscode-ext/src/api.ts [hunk 4]
```diff
@@ -0,0 +1,267 @@
+import * as vscode from 'vscode';
+import * as path from 'node:path';
+import * as fs from 'node:fs';
+import { ChildProcessWithoutNullStreams, spawn } from 'child_process';
+import { globalContext } from './extension';
+const util = require('node:util');
+const exec = util.promisify(require('node:child_process').exec);
```
Comment (Razz4780) [inline vscode-ext/src/api.ts:7]: These two should no longer be needed, right?

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: vscode-ext/src/api.ts (hunk 4 above)
Comment (infiniteregrets) [reply vscode-ext/src/api.ts:7]: yes you are right, thanks!

**Review by Razz4780 — APPROVED** [review]

**Review by eyalb181 — APPROVED** [review]
